tests/testlib/s2n_mem_testlib.c

```diff
@@ -0,0 +1,169 @@
+#include "utils/s2n_mem.c"
+
+struct s2n_mem_test_cb_ctx {
+    struct s2n_stuffer mallocs;
+} s2n_mem_test_ctx = { 0 };
+
+static s2n_mem_malloc_callback s2n_mem_malloc_cb_backup = NULL;
+static s2n_mem_free_callback s2n_mem_free_cb_backup = NULL;
+
+static int s2n_mem_test_malloc_cb(void **ptr, uint32_t requested, uint32_t *allocated);
+static int s2n_mem_test_free_cb(void *ptr, uint32_t size);
+
+static S2N_RESULT s2n_mem_test_set_callbacks()
+{
+    if (s2n_mem_malloc_cb != s2n_mem_test_malloc_cb) {
+        s2n_mem_malloc_cb_backup = s2n_mem_malloc_cb;
+        s2n_mem_malloc_cb = s2n_mem_test_malloc_cb;
+    }
+    if (s2n_mem_free_cb != s2n_mem_test_free_cb) {
+        s2n_mem_free_cb_backup = s2n_mem_free_cb;
+        s2n_mem_free_cb = s2n_mem_test_free_cb;
+    }
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_mem_test_unset_callbacks()
+{
+    if (s2n_mem_malloc_cb_backup != NULL) {
+        s2n_mem_malloc_cb = s2n_mem_malloc_cb_backup;
+    }
+    if (s2n_mem_free_cb_backup != NULL) {
+        s2n_mem_free_cb = s2n_mem_free_cb_backup;
+    }
+    return S2N_RESULT_OK;
+}
+
+static int s2n_mem_test_malloc_cb(void **ptr, uint32_t requested, uint32_t *allocated)
+{
+    int result = s2n_mem_malloc_cb_backup(ptr, requested, allocated);
+    POSIX_GUARD(result);
+
+    struct s2n_mem_test_malloc new_info = {
+        .requested = requested,
+        .ptr = *ptr,
+        .allocated = *allocated,
+        .freed = false,
+    };
+
+    POSIX_GUARD_RESULT(s2n_mem_test_unset_callbacks());
+    POSIX_GUARD(s2n_stuffer_write_bytes(&s2n_mem_test_ctx.mallocs,
+            (uint8_t *) &new_info, sizeof(new_info)));
+    POSIX_GUARD_RESULT(s2n_mem_test_set_callbacks());
+
+    return result;
+}
+
+static int s2n_mem_test_free_cb(void *ptr, uint32_t size)
+{
+    int result = s2n_mem_free_cb_backup(ptr, size);
+    POSIX_GUARD(result);
+
+    struct s2n_stuffer read_copy = s2n_mem_test_ctx.mallocs;
+    while (s2n_stuffer_data_available(&read_copy)) {
+        uint8_t *mem = s2n_stuffer_raw_read(&read_copy, sizeof(struct s2n_mem_test_malloc));
+        struct s2n_mem_test_malloc *info = (struct s2n_mem_test_malloc *) (void *) mem;
+        if (info->ptr == ptr) {
+            info->freed = true;
+        }
+    }
+    return result;
+}
+
+S2N_RESULT s2n_mem_test_init_callbacks(void *ctx)
+{
+    RESULT_GUARD(s2n_mem_test_free_callbacks(ctx));
+    RESULT_GUARD_POSIX(s2n_stuffer_growable_alloc(&s2n_mem_test_ctx.mallocs, 0));
+    RESULT_GUARD(s2n_mem_test_set_callbacks());
+    return S2N_RESULT_OK;
+}
+
+S2N_CLEANUP_RESULT s2n_mem_test_free_callbacks(void *ctx)
+{
+    (void) ctx;
+    RESULT_GUARD(s2n_mem_test_unset_callbacks());
+    RESULT_GUARD_POSIX(s2n_stuffer_free(&s2n_mem_test_ctx.mallocs));
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_mem_test_wipe_callbacks()
+{
+    RESULT_GUARD_POSIX(s2n_stuffer_wipe(&s2n_mem_test_ctx.mallocs));
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_mem_test_get_malloc_count(uint32_t *count)
```

Comment:
This is a large amount of test functions that aren't actually being used yet. I worry about putting quite so many in and then never using them? Like, the benefit of having them in this PR is that I'm certain that these methods work as intended, since you tested them all. But like, will we actually use _all_ of these test functions?

Reply:
I made the getters static. They were awkward to use in tests anyway, which is why I created the asserts. You're right, if we eventually need them we can always add them back.

For the others, I already have a use case for them: https://github.com/lrstewart/s2n/blob/d65447204a2fde2d5bdd8ddff26ee48435ad08f8/tests/unit/s2n_ktls_io_test.c#L980-L984